README.md
[![](https://ptkdev.it/img/bot/instagram-bot-lib.png)](https://instagram.bot.ptkdev.io)

# Instagram Bot (Library)

[![](https://img.shields.io/badge/license-GLPv3-brightgreen.svg)](#) [![](https://img.shields.io/badge/powered%20by-puppeteer-46aef7.svg)](https://github.com/GoogleChrome/puppeteer) [![](https://img.shields.io/badge/version-v0.9.10-lightgrey.svg)](https://github.com/social-manager-tools/instagram-bot-lib/releases) [![](https://img.shields.io/badge/chat%20on-slack-orange.svg)](http://slack.ptkdev.io) [![](https://img.shields.io/badge/chat%20on-discord-7289da.svg)](http://discord.ptkdev.io) [![](https://img.shields.io/badge/blog-medium-2AE176.svg)](http://blog.ptkdev.io) [![](https://img.shields.io/badge/twitter-ptkdevio-2AA3EF.svg)](https://twitter.com/ptkdevio)

[![](https://img.shields.io/badge/donate-patreon-F87668.svg)](http://patreon.ptkdev.io) [![](https://img.shields.io/badge/donate-paypal-46AFE0.svg)](http://paypal.ptkdev.io) [![](https://img.shields.io/badge/buy%20me-coffee-4B788C.svg)](http://coffee.ptkdev.io) [![](https://img.shields.io/badge/help-support@ptkdev.io-fbbc05.svg)](mailto:support@ptkdev.io)

[![](https://ptkdev.it/img/bot/ptkdev-instagram-bot.gif)](https://instagram.bot.ptkdev.io)

## What it does
This library (npm module) provides api for include [instagram-bot.js](https://github.com/social-manager-tools/instagram-bot.js) functions in your application. This bot helps you increase the engagement level of your Instagram profile through different social algorithms. Increase the likes on your photos and followers!

## Features
* [✓] Easy to use
* [✓] Login
* [✓] 2FA (bad location)
* [✓] 2FA (sms pin enabled)
* [✓] Multi-Session
* [✓] Multi-Platform (Windows 10, Mac OSX, Linux, [Raspberry PI 3](https://github.com/social-manager-tools/instagram-bot.js/blob/master/INSTALL.md))
* [✓] Error management feature (bad pin, bad password)
* [✓] Screenshot and Verbose logger
* [✓] Like Mode Classic: bot selects a random hashtag from a config list and likes 1 random photo, and can repeat this all time.
* [✓] Like Mode Realistic: bot selects a random hashtag from a config list and likes fast 10-12 photos, it sleeps 15-20min and repeats this all time. Sleeps at night.
* [✓] Like Mode Competitor Users: it selects an account, selects random followers, likes 10-12 photo and sleeps 15-20min. Sleeps at night.
* [✓] Like Mode Superlike: it selects random hashtag from a config list and likes 3 random photos of the same user.
* [✓] Comment Mode Classic: it selects random comments and random hashtags and writes comments under photos.
* [✓] Follow/Defollow Classic: follow 30 users, defollow first and rotate (in loop). This method is not detect from socialblade. ~1h | 300 follow-defollow/day.

## Fast usage
1. Run `npm install instagrambotlib`
2. Get [config.js](https://raw.githubusercontent.com/social-manager-tools/instagram-bot-lib/master/config.js.tpl) file for step 3, fill it properly and remove `.tpl` suffix.
3. On your code require library and run bot, example:
```
    const config = require ("./config");
    const Bot = require("instagrambotlib");
    let bot = new Bot(config);
    bot.start();
```
4. If work add star :star: at this project :heart:
5. If you want help me: **[donate on paypal](http://paypal.ptkdev.io)** or become a **[backer on patreon](http://patreon.ptkdev.io)**.

For advanced configuration see [INSTALL.md](https://github.com/social-manager-tools/instagram-bot-lib/blob/master/INSTALL.md).

#### 2FA: SMS Pin
If you received sms or email pin edit `loginpin.txt` and insert it on first line. Wait 50-60 seconds...

#### Tips: hide browser
Edit `config.js` and switch `chrome_headless` option to `true`.

#### Check if work:
See images in ./logs/screenshot or disable `chrome_headless` flag.

## API 
Method | Description
------------ | -------------
start() | run instagram-bot.js
stop()  | stop instagram-bot.js

## Bugs
1. `[ERROR] login: The username you entered doesn't belong to an account. Please check your username and try again. (restart bot and retry...)`
* Why happen? Instagram desktop is in overcapacity. Happen at 12-14 and 19-21 all days. 
* Solution: Login in other time or Logout from your instagram app, and login again. Reboot bot and retry... Try and retry, and retry, and retry... Or stop bot and wait 2-3h...

2. `Error: Protocol error (Page.captureScreenshot): Target closed.`
* Why happen? macOS don't support correctly screenshot from puppeteer
* Solution: set `screenshot` on `false` in `config.js`

3. `This code is no longer valid. Please request a new one. (400) (/accounts/login/ajax/two_factor/)` 
* Why happen? Instagram bug at login
* Solution: disable at moment 2FA or try old version of chrome (edit `config.js` set `executable_path`)

## Desktop setup (GUI Version)
1. [Social Manager Tools GUI](https://socialmanagertools.ptkdev.io/).
2. Run application.

## Docker setup
If you prefer to run this using Docker, an official container is available from the [Docker Hub](https://hub.docker.com/r/socialmanagertools/instagram-bot.js).

In order to run it, copy the `config.js.tpl` file, configure it as you prefer, then use it through volume mapping,
like in this example:

```sh
docker run --restart=always --name=instagram-bot -d -v /path/to/config.js:/app/configs/config.js socialmanagertools/instagram-bot.js:amd64
```

**AVAILABLE TAGS:** `amd64` (64bit), `i386` (32bit),`armv7` (Raspberry PI 2), `armv8` (Raspberry PI 3)

**WARNING:** with docker is mandatory edit `config.js` and set `chrome_headless` on `true` and set `executable_path` to `/usr/bin/chromium-browser`. Without this fix docker don't work.

## Roadmap
See full roadmap (open task, todo and bugs) in [project page](https://github.com/social-manager-tools/instagram-bot.js/projects?query=is%3Aopen+sort%3Aname-asc).
* [v0.9.X](https://github.com/social-manager-tools/instagram-bot.js/projects/3)
* [v1.0.X](https://github.com/social-manager-tools/instagram-bot.js/projects/3)

## Sorry for snake_case
I love :snake: snake_case syntax sorry for this :sob: don't hate me.

[![](https://socialmanagertools.ptkdev.io/img/socialmanagertools_logo.png)](https://github.com/social-manager-tools)

# Social Manager Tools

[Social Manager Tools GUI](https://github.com/social-manager-tools/social-manager-tools)  
[InstagramBot.js](https://github.com/social-manager-tools/instagram-bot.js) ([LIB](https://github.com/social-manager-tools/instagram-bot-lib))  
[TwitterBot.js](https://github.com/social-manager-tools/twitter-bot.js) ([LIB](https://github.com/social-manager-tools/twitter-bot-lib))  
[FacebookBot.js](https://github.com/social-manager-tools/facebook-bot.js) ([LIB](https://github.com/social-manager-tools/facebook-bot-lib))  
[WordpressTelegramBot.js](https://github.com/social-manager-tools/wordpress-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/wordpress-telegram-bot-lib))  
[MediumTelegramBot.js](https://github.com/social-manager-tools/medium-telegram-bot.js) ([LIB](https://github.com/social-manager-tools/medium-telegram-bot-lib))  

# License

GNU GENERAL PUBLIC LICENSE

Copyright (c) 2018 Patryk Rzucidło (PTKDev)


CHANGELOG.md
# v0.9.10 (TBD, 2018)
* Developers: [Guidelines](https://github.com/social-manager-tools/instagram-bot-lib/blob/nightly/DEV_GUIDELINES.md) for contributing at project.
* Test: add mocha and chai
* Upgrade: puppeteer v1.7.0


# v0.9.9 (16 October, 2018)
* Fix: fdfmode_classic #42 (again again...)
* Fix: 2FA (Location)
* Fix: Close button (ui)
* Fix: lose connection bot stop, now retry work.


# v0.9.8 (14 October, 2018)
* Fix: fdfmode_classic #42 (again)
* Fix: 2FA doesn't work #48 (after 3 months... thanks instagram!)
* Update: default value of `chrome_headless` is `true`
* Update: all library of package.json


# v0.9.7 (08 October, 2018)
* Fix: fdfmode_classic #42


# v0.9.6 (05 October, 2018)
* Fix: login random don't work


# v0.9.5 (02 October, 2018)
* Fix: likemode_classic new instagram selectors
* Fix: likemode_realistic new instagram selectors
* Fix: likemode_superlike new instagram selectors
* Fix: likemode_competitor_users new instagram selectors
* Fix: comment_mode new instagram selectors
* Fix: fdfmode_classic new instagram selectors
* API: Fix stop() again, now really work.


# v0.9.4 (26 September, 2018)
* Fix: API::stop() now work correctly
* Fix: Follow/Defollow Mode Realistic random userpage is 404
* Fix: docker container (by [Julian Xhokaxhiu](https://github.com/julianxhokaxhiu)) #95
* Upgrade: puppeteer v1.6.2


# v0.9.3 (22 September, 2018)
* Fix: Follow/Defollow Mode Realistic stop working if photo or account are removed
* Fix: Like Mode Realistic correct like/don't like message
* Performance: all actions are now fastest than 1-2 seconds


# v0.9.2 (15 September, 2018)
* Downgrade: puppeteer v1.4.0 (#27)


# v0.9.1 (15 September, 2018)
* Fix: Node is either not visible or not an HTMLElement (#82 #86 #84)
* Fix: Random crash of Follow/Defollow Mode Realistic
* Refactor: 2FA flow
* Update: New user-agent
* Upgrade: puppeteer v1.8.0


# v0.9.0b (09 September, 2018)
* Fix: bad creation of folders `databases` and `logs` (#79)


# v0.9.0 (08 September, 2018)
* Feature: Follow/Defollow Mode (#77)
* Feature: Support sqllite connector (#70)
* Fix: Like Realistic random freeze
* Fix: Comment Mode don't work
* Fix: Like Competitor Mode (#76)
* API: stop() | stop the bot from your application
* Update: bot_fastlike_min and max from config.js now all mode use bot_likeday_min and max
* Update: bot_sleep_night disabled at default
* Upgrade: puppeteer v1.7.0


# v0.8.2 (26 July, 2018)
* Fix: default path of logs in `config.js.tpl`


# v0.8.1 (25 July, 2018)
* Fix: Unliking already liked pictures (all mode) [Bug: #59]
* Fix: Commenting already commented pictures
* Update: `likemode_realistic` now is default mode in `config.js.tpl`
* Update: `900` now is default value of max like/day in `config.js.tpl`


# v0.8.0 (17 July, 2018)
* Fix: path of logs and screenshot


# v0.7.5c and v0.7.5d (11 July, 2018)
* Fix: likemode_superlike timeout error


# v0.7.5b (09 July, 2018)
* Feature: path log/pin and uifix() for social-manager-tools app


# v0.7.5 (03 July, 2018)
* Update: Porting of instagram-bot.js v0.7.5


# v0.7.4c (29 June, 2018)
* Fix: Porting of instagram-bot.js v0.7.4


# v0.7.4b (28 June, 2018)
* Update: Porting of instagram-bot.js v0.7.4


# v0.6.1 (20 February, 2018)
* Update: Porting of instagram-bot.js v0.6.1

modules/common/utils.js
/**
 * Utils
 * =====================
 * Logger and other functions...
 *
 * @author:     Patryk Rzucidlo [@ptkdev] <support@ptkdev.io> (https://ptkdev.it)
 * @license:    This code and contributions have 'GNU General Public License v3'
 *
 */
require("colors");
class Utils {
    constructor (bot, browser, config) {
        this.bot = bot;
        this.browser = browser;
        this.config = config;
        this.fs = require("fs");
        this.LOG_NAME = "utils";
        this.LOG = require("../logger/types");
        this.MAP_COLORS = require("../logger/types").MAP_COLORS;
        this.Log = require("../logger/log");
        this.log = new this.Log(this.LOG_NAME, this.config);
    }

    /**
     * Donate
     * =====================
     * Patreon link
     *
     */
    donate () {
        this.log.warning("Bot work? Please donate for support this project!");
        this.log.warning("Donate with patreon: http://patreon.ptkdev.io");
        this.log.warning("Donate with paypal: http://paypal.ptkdev.io");
    }

    /**
     * Compare package version
     * =====================
     * Compare current version with passed parameter
     * @param {string} version to compare with
     * @return {string} 0 if same version, 1 if older, -1 if newer
     *
     */
    compare_version (version) {
        const current_version = require("../../version").version;
        const compare_versions = require("compare-versions");
        return compare_versions(current_version, version);
    }

    /**
     * Check updates
     * =====================
     * Bot is updated? Yes/no
     *
     */
    check_updates (version) {
        let request = require("request");
        let log = this.log;
        request.get("https://api.ptkdev.io/v1/bot/instagram/version/", function (err, res, last_release) {
            if (err) {
                log.error("Is impossible contact api.ptkdev.io server, wifi is on?");
            } else {
                if (version !== last_release) {
                    log.warning(`Bot release v${last_release} available! Current version: v${version}`);
                } else {
                    log.info("Bot is updated! :D");
                }
            }
        });
    }

    /**
     * Create empty files: loginpin.txt, logs amd db
     * =====================
     * This fix errors at boot
     *
     */
    async create_write_file (name, path, content) {
        let self = this;
        return new Promise(function (resolve_write, reject_write) {
            self.fs.writeFile(path, content, function (err) {
                if (err) {
                    self.log.error(`${name} don't created: ${err}`);
                    reject_write(err);
                } else {
                    self.log.info(`${name} created`);
                }
                resolve_write(true);
            });
        });
    }

    async create_is_exist (name, path) {
        let self = this;
        return new Promise(function (resolve_exists) {
            self.fs.open(path, "wx", function (exists) {
                if (exists && exists.code === "EEXIST") {
                    self.log.info(`${name} exist`);
                    resolve_exists("exist");
                } else {
                    resolve_exists("create");
                }
            });
        });
    }

    async create_empty (name, path, content) {
        if (await this.create_is_exist(name, path) == "create") {
            await this.create_write_file(name, path, content);
        }
    }

    /**
     * Init all emptys filesthis.
     * =====================
     * This fix errors at boot
     *
     */
    async init_empty () {
        await this.create_empty("loginpin.txt", this.config.pin_path, "123456");

        await this.create_empty("logs/debug.log", this.config.log_path, "");
        await this.create_empty("logs/errors.log", this.config.logerr_path, "");

        await this.create_empty("databases/fdf.db", this.config.fdfdatabase_path, "");
        await this.create_empty("databases/logs.db", this.config.logdb_path, "");
    }

    compute_interval_between_run_in_seconds (target_count_per_day) {
        if (target_count_per_day <= 0) {
            throw new Error("target_count_per_day must be positive");
        }
        return Math.ceil((24 / target_count_per_day) * 11 *60);
    }

    /**
     * Screenshot
     * =====================
     * Save screenshot from chrome
     *
     */
    async screenshot (func, name) {
        if (this.config.log.screenshot) {
            try {
                await this.bot.screenshot({path: `${this.config.screenshot_path + this.config.instagram_username}_${func}_${name}.jpg`});
                this.log.info("Cheese! Screenshot!");
            } catch (err) {
                this.log.error(`screenshot: error ${err}`);
            }
        }
    }

    /**
     * Random
     * =====================
     * Random number between two numbers
     *
     */
    random_interval (min, max) {
        return (Math.floor(Math.random() * (max - min + 1)) + min) * 1000;
    }

    /**
     * Kill Process if browser page is closed
     * =====================
     * Exit from node
     *
     */
    async keep_alive () {
        let pages = null;

        try {
            pages = (await this.browser.pages()).map(t => t.url());
        } catch (err) {
            this.log.info("Bye bye! Shutdown... wait ~30sec for the bot stopping...");
            return false;
        }

        if (pages.length == 2) {
            return true;
        } else {
            
            this.log.info("Bye bye! Shutdown... wait ~30sec for the bot stopping...");
        }

        return false;
    }

    /**
     * Get random number between two numbers
     * @param min
     * @param max
     * @return {number}
     */
    random_number () {
        return (Math.floor(Math.random() * (20 - 10 + 1)) + 10);
    }

    /**
     * Mix array element
     * @param arr
     * @return array
     */
    mix_array (arr) {
        return arr.sort(function () {
            return 0.5 - Math.random();
        });
    }

    /**
     * Sleep
     * =====================
     * Zzz
     *
     */
    sleep (sec) {
        return new Promise(resolve => setTimeout(resolve, sec));
    }

    /**
     * Check is debug
     * @return {boolean}
     */
    is_debug () {
        return this.config.debug === true;
    }

    /**
     * Get random hash tag from config file
     * @return {string}
     */
    get_random_hash_tag () {
        return this.config.instagram_hashtag[Math.floor(Math.random() * this.config.instagram_hashtag.length)];
    }

    /**
     * Fix ui configs params
     * =====================
     * String aparms to booleans from social-manager-tools
     *
     */
    fixui (config) {
        config.bot_likeday_min = parseInt(config.bot_likeday_min);
        config.bot_likeday_max = parseInt(config.bot_likeday_max);
        config.bot_superlike_n = parseInt(config.bot_superlike_n);
        if (config.chrome_headless == "enabled") {
            config.chrome_headless = false;
        } else {
            config.chrome_headless = true;
        }

        return config;
    }

    /**
     * Default config.js
     * =====================
     * Get default value if config.js is not updated from config.js.tpl
     *
     */
    fixconfig (config) {
        if (typeof config.debug === "undefined") {
            config.debug = true;
            this.log.error("config.debug use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.login === "undefined") {
            config.login = true;
            this.log.error("config.login use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.instagram_username === "undefined") {
            config.instagram_username = "ptkdev";
            this.log.error("config.instagram_username use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.instagram_password === "undefined") {
            config.instagram_password = "password";
            this.log.error("config.instagram_password use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.instagram_hashtag === "undefined") {
            config.instagram_hashtag = ["follow"];
            this.log.error("config.instagram_hashtag use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.instagram_pin === "undefined") {
            config.instagram_pin = "sms";
            this.log.error("config.instagram_pin use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_mode === "undefined") {
            config.bot_mode = "likemode_classic";
            this.log.error("config.bot_mode use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_likeday_min === "undefined") {
            config.bot_likeday_min = 300;
            this.log.error("config.bot_likeday_min use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_likeday_max === "undefined") {
            config.bot_likeday_max = 600;
            this.log.error("config.bot_likeday_max use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_sleep_night === "undefined" || (config.bot_sleep_night !== true && config.bot_sleep_night !== false)) {
            config.bot_sleep_night = true;
            this.log.error("config.bot_sleep_night use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_start_sleep === "undefined") {
            config.bot_start_sleep = "7:00";
            this.log.error("config.bot_start_sleep use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_superlike_n === "undefined") {
            config.bot_superlike_n = 3;
            this.log.error("config.bot_superlike_n use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_followday === "undefined") {
            config.bot_followday = 300;
            this.log.error("config.bot_followday use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.bot_userwhitelist === "undefined") {
            config.bot_userwhitelist = [""];
            this.log.error("config.bot_userwhitelist use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.chrome_headless === "undefined") {
            config.chrome_headless = false;
            this.log.error("config.chrome_headless use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.chrome_options === "undefined") {
            config.chrome_options = ["--disable-gpu", "--no-sandbox", "--window-size=1920x1080"];
            this.log.error("config.chrome_options use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.executable_path === "undefined") {
            config.executable_path = "";
            this.log.error("config.executable_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.screenshot_path === "undefined") {
            config.screenshot_path = "./logs/screenshot/";
            this.log.error("config.screenshot_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.logerr_path === "undefined") {
            config.logerr_path = "./logs/errors.log";
            this.log.error("config.logerr_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.log_path === "undefined") {
            config.log_path = "./logs/debug.log";
            this.log.error("config.log_path use the default value, update your config.js from config.js.tpl");
        }
        if (typeof config.pin_path === "undefined") {
            config.pin_path = "./loginpin.txt";
            this.log.error("config.pin_path use the default value, update your config.js from config.js.tpl");
        }

        config.bot_fastlike_max = this.compute_interval_between_run_in_seconds(config.bot_likeday_min);
        config.bot_fastlike_min = this.compute_interval_between_run_in_seconds(config.bot_likeday_max);

        config.bot_fastlikefdf_max = this.compute_interval_between_run_in_seconds(config.bot_followday - 30);
        config.bot_fastlikefdf_min = this.compute_interval_between_run_in_seconds(config.bot_followday);

        return config;
    }
}

module.exports = (bot, browser, config) => {
    return new Utils(bot, browser, config);
};